Recognize Skip/Take chains on lazy sequences.

jamesqo opened the pull request:
I added a new `EnumerablePartition<TSource>` class that lets us now recognize patterns like `Take.Skip`, or vice versa, for lazy sequences. We also remove a layer of indirection from some common operations following these methods, e.g. `Take.ToArray`, `Skip.ToArray`, etc.

Perf results can be found [here](https://gist.github.com/jamesqo/07eb6b38d69431d7270900e3a7519c32); test code [here](https://github.com/jamesqo/Dotnet/blob/502e3c302575d04824e7337bec4d635abc8dc548/Program.cs). There is a ~20% speedup that results from removing an indirection layer from the chain.

cc @JonHanna, @VSadov, @stephentoub

jamesqo commented on src/System.Linq/src/System/Linq/Partition.cs:
```diff
@@ -389,5 +389,245 @@ public int GetCount(bool onlyIfCheap)
+                    if (SkipBeforeFirst(en) && en.MoveNext())
+                    {
+                        var builder = new LargeArrayBuilder<TSource>(initialize: true);
+                        int toTake = _maxIndexInclusive - _minIndexInclusive + 1;
```
Note: This can overflow if `_maxIndexInclusive` is int.MaxValue, and `_minIndexInclusive` is 0 (which happens for `e.Skip(0)`). It doesn't matter though, since we decrement the value right away.

stephentoub replied:
Please add a comment about that.

jamesqo commented on src/System.Linq/src/System/Linq/Partition.cs:
```diff
@@ -389,5 +389,245 @@ public int GetCount(bool onlyIfCheap)
+                return (uint)maxIndex >= (uint)_maxIndexInclusive ? this : new EnumerablePartition<TSource>(_source, _minIndexInclusive, maxIndex);
+            }
+
+            public TSource TryGetElementAt(int index, out bool found)
```
I would have avoided optimizing the `TryGet*` functions if I could have, since they take a lot of code and are not as valuable to optimize as the other functions. But, currently they come with implementing `IPartition`.

JonHanna commented:
The combination `.Skip(…).Take(…)` is a common one, so this seems a reasonable case to cover in this way.

stephentoub commented on src/System.Linq/tests/TakeTests.cs:
```diff
@@ -450,5 +450,16 @@ public void RepeatEnumeratingNotList()
+        [Theory]
+        [InlineData(1000)]
+        [InlineData(1000000)]
+        [InlineData(int.MaxValue)]
```
This test is failing.

jamesqo replied:
Fixed

stephentoub commented on src/System.Linq/src/System/Linq/Partition.cs:
```diff
@@ -389,5 +389,245 @@ public int GetCount(bool onlyIfCheap)
+                        int taken = _state - 3;
+                        if (taken <= _maxIndexInclusive - _minIndexInclusive && _enumerator.MoveNext())
+                        {
+                            _state++;
```
Doesn't using _state like this cause a problem for really long enumerables with > int.MaxValue - 3 elements?

jamesqo replied:
Will move to long.

stephentoub commented on src/System.Linq/src/System/Linq/Partition.cs:
```diff
@@ -389,5 +389,245 @@ public int GetCount(bool onlyIfCheap)
+                {
+                    using (IEnumerator<TSource> en = _source.GetEnumerator())
+                    {
+                        if (SkipBeforeFirst(en) && TakeBefore(index, en) && en.MoveNext())
```
Please make sure tests are added to cover all of these various branches, here and elsewhere.

stephentoub commented on src/System.Linq/src/System/Linq/Partition.cs:
```diff
@@ -389,5 +389,245 @@ public int GetCount(bool onlyIfCheap)
+
+                // Did we finish skipping before we reached the end of the sequence?
+                return toSkip == 0;
+            }
```
Can this whole method just be:
```C#
private bool SkipBeforeFirst(IEnumerator<TSource> en) => TakeBefore(_minIndexInclusive, en);
```
?

jamesqo replied:
@stephentoub Good idea. In fact, I renamed the method to be `SkipBefore` and had `TryGetElementAt` do `SkipBefore(_minIndexInclusive + index, en)`. Should be slightly more efficient.

stephentoub commented on src/System.Linq/src/System/Linq/Skip.cs:
```diff
@@ -41,26 +41,7 @@ public static IEnumerable<TSource> Skip<TSource>(this IEnumerable<TSource> sourc
-                    }
-                }
-            }
+            return new EnumerablePartition<TSource>(source, count, int.MaxValue);
```
Doesn't this end up limiting enumerables to those with int.MaxValue elements?  That limitation was not there before.

jamesqo replied:
I can change this to use `long` instead. It's unlikely that an enumerable that comes close to `long.MaxValue` elements can be iterated in a reasonable amount of time.

jamesqo commented:
Marking as no merge temporarily. Since long fields are now being used, which takes up an extra 16 bytes, I'll have to re-measure the memory impact this change has. (Also, there's some more refactoring I'd like to do.)

jamesqo commented:
@stephentoub Re. the issue with enumerables larger than `int.MaxValue`-- to support enumerables longer than that, I changed the type of the index fields to `long`, and added another `long` field, `_taken`, since we can no longer use `_state - 3` since `_state` is an int. However, this causes the field size to increase by 16 bytes. I made a [test app](https://gist.github.com/jamesqo/25e02482e5a9257b52ee3b6c91bfc556) where `Skip` was called 1m times in a loop, and it looks like the # of GCs increases by 1-3 out of 20.

Obviously this is no good; what do you propose I do? Would it be OK if I changed the type of the fields to `uint` instead, so we only support enumerables up to `uint.MaxValue` (minus around 3 or 4)?

jamesqo commented:
@stephentoub, I managed to refactor the iterator to support enumerables > int.MaxValue, while still using `int` for the indices. The trick is `_maxIndexInclusive` is set to -1 now if we're taking an indeterminate number of items (e.g. all Skips). I've also added tests and done some refactoring.

@JonHanna, would appreciate if you could review as well. I ran into a couple of subtle integer overflow bugs whilst working on this PR, so the more eyes the better.

jamesqo commented on src/System.Linq/src/System/Linq/Partition.cs:
```diff
@@ -389,5 +389,308 @@ public int GetCount(bool onlyIfCheap)
+                            remaining--;
+                            builder.Add(en.Current);
+                        }
+                        while (remaining >= comparand && en.MoveNext());
```
We're dishing out an extra unnecessary write/branch per iteration for the loop if `HasLimit` is false (e.g. if only `Skip` has been called, and not `Take`). It could be specialized like this:

```cs
if (!HasLimit)
{
    do { builder.Add(en.Current); } while (en.MoveNext());
}
else
{
    do { remaining--; builder.Add(en.Current); } while (remaining >= 0 && en.MoveNext());
}
```

I figured it wasn't worth the extra code though.

jamesqo commented on src/System.Linq/src/System/Linq/Partition.cs:
```diff
@@ -389,5 +389,308 @@ public int GetCount(bool onlyIfCheap)
+                return (int)SkipAndCount((uint)index, en);
+            }
+
+            private static uint SkipAndCount(uint index, IEnumerator<TSource> en)
```
Had to add a `uint` overload of this function to support scenarios where `_maxIndexInclusive` is exactly int.MaxValue & we call `Count()`, for example `e.Skip(1).Take(int.MaxValue).Count()`. We don't lose any validation at callsites other than `GetCount` though, since the signed wrapper asserts.

jamesqo commented on src/System.Linq/src/System/Linq/Partition.cs:
```diff
@@ -389,5 +389,308 @@ public int GetCount(bool onlyIfCheap)
+        private sealed class EnumerablePartition<TSource> : Iterator<TSource>, IPartition<TSource>
+        {
+            private readonly IEnumerable<TSource> _source;
+            private readonly int _minIndexInclusive;
```
@VSadov I originally had something like that, actually, using ints. But since all of the other partitions used `_minIndexInclusive` and `_maxIndexInclusive` I decided to follow their lead.

> Why not uint _start and uint _length?

I used ints for the fields, since it's more idiomatic. If we know the length (`Take` has been called atl. once) then the count will fit in an int anyway. Also, it'd still be possible that `_start` could overflow during a chained `Skip` / `Take` & we'd have to wrap ourselves in another iterator.

jamesqo commented:
@stephentoub I updated this PR to match the changes in #14020. Can you please review when you have time? Thanks.

VSadov commented:
LGTM
@stephentoub  - could you take another look at the changes?

stephentoub commented on src/System.Linq/src/System/Linq/Partition.cs:
```diff
@@ -389,5 +389,310 @@ public int GetCount(bool onlyIfCheap)
+            // on how many elements we can have.
+            private bool HasLimit => _maxIndexInclusive != -1;
+
+            private int Limit => (_maxIndexInclusive + 1) - _minIndexInclusive; // This is that upper bound.
```
This value is only meaningful if HasLimit is true, right?  Should we add a ```Debug.Assert(HasLimit)``` to the getter?

stephentoub commented on src/System.Linq/src/System/Linq/Partition.cs:
```diff
@@ -389,5 +389,310 @@ public int GetCount(bool onlyIfCheap)
+                            if (HasLimit)
+                            {
+                                // If we are taking an unknown number of elements, it's important not to increment _state.
+                                // _state - 3 may eventually end up overflowing & we'll hit the Dispose branch even though
```
How expensive would it be to have an outerloop test that overflowed to test this condition?

jamesqo replied:
Well it would be 2B * 3 virtual method calls (`MoveNext`, `Current`, and `MoveNext` on the iterator) plus all of the other logic, so yes, probably too expensive for a test.

stephentoub replied:
Yeah.  In a few places in our tests we do use reflection to put the object into a state that would make it hit such cases quickly, e.g.
https://github.com/dotnet/corefx/blob/master/src/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/HandleCollectorTests.cs#L63
Would it be worth doing something like that here?  Or we don't think there's enough risk to warrant that kind of fragility?

jamesqo replied:
> Would it be worth doing something like that here? Or we don't think there's enough risk to warrant that kind of fragility?

It is low risk, but at the same time I think it is unlikely the field name of `_state` will change. I will add a test.

jamesqo commented on src/System.Linq/src/System/Linq/Partition.cs:
```diff
@@ -389,5 +389,310 @@ public int GetCount(bool onlyIfCheap)
+            {
+                if (onlyIfCheap)
+                {
+                    return -1;
```
Will add one.

jamesqo replied:
@stephentoub Made the change. Something to note: While adding coverage for this, I also realized I had forgotten to override `Select` in this type. ([`Select`](https://github.com/dotnet/corefx/blob/master/src/System.Linq/src/System/Linq/Select.cs#L25) checks for an Iterator first, then an IPartition, and the default implementation of `Iterator.Select` is to return a `SelectEnumerableIterator` rather than a `SelectIPartitionIterator`.) So thank you for pointing that out.

stephentoub commented:
Thanks, @jamesqo. LGTM.
